### PR TITLE
ENH: pcds-5.4.2 and script updates

### LINKED ISSUE #243: Add deploy script for cleaning up old hard drive conda envs
A few of the hosts are starting to fill up on disk space from the on-disk conda releases

We need to write a script that:
- checks what version the hutch is running
- keeps the most recent versions and n versions prior to the active version
- deletes everything else older

for some n that we should agree on

### LINKED ISSUE #245: Addition of Qt debug tool GammaRay (dev? prod?)
## What is it?

Documentation: https://doc.qt.io/GammaRay/index.html

> [GammaRay](http://doc.qt.io/GammaRay) is an introspection tool for Qt applications developed by [KDAB](https://www.kdab.com/). Leveraging the [QObject](http://doc.qt.io/qt-5/qobject.html) introspection mechanism it allows you to observe and manipulate your application at runtime. This works both locally on your workstation and remotely on an embedded target.

## Feedstock

https://github.com/conda-forge/gammaray-feedstock

## Possible issue

Adding it to the production environment may be problematic once we reach Qt 5.15 as the feedstock build is failing. Based on the two unmerged PRs in that repository, it does not appear that the repository is actively maintained.

So - possibly this could be a dev-env (or temporary prod) addition while it works with our installed Qt version.

### LINKED ISSUE #246: Add paramiko
https://www.paramiko.org/

Required for the switchtool that Mike Browne is working on

### LINKED ISSUE #247: Add Flask
Required for running the database server for the in-progress pmps database

### LINKED ISSUE #248: Identify "interactive-only" packages to freeze versions of
Packages like IPython that primarily present themselves as interactive interfaces are difficult to test automatically for compatibility with user operations at SLAC/LCLS. Bugs and new behavior easily slip through unnoticed as the library continues to improve its user experience.

Need to:
- lock the version of IPython and upgrade only very deliberately
- identify other packages that may need the same treatment

### PR DESCRIPTION
**Pull request by ZLLentz**
## Overview
- Package update that does as small a change as possible
- closes #243
- closes #245
- closes #246
- closes #247
- closes #248

## Scripts
- `clean_old_packs.py` to help clean up unused conda-pack environments on operator console disks
- `create_incremental_env.sh` to create an environment as similar as possible to a previous environment but with our new package specs

## CI
- Add a ci build for the incremental script

## Package Notes
PCDS Package Updates
--------------------

|   Package    |  Old   |  New   |                        Release Notes                         |
|:------------:|:------:|:------:|:------------------------------------------------------------:|
|    happi     | 1.13.0 | 1.14.0 |    https://github.com/pcdshub/happi/releases/tag/v1.14.0     |
| hutch-python | 1.16.0 | 1.17.0 | https://github.com/pcdshub/hutch-python/releases/tag/v1.17.0 |
|   pcdscalc   | 0.3.2  | 0.3.3  |   https://github.com/pcdshub/pcdscalc/releases/tag/v0.3.3    |
| pcdsdevices  | 6.1.0  | 6.3.0  |  https://github.com/pcdshub/pcdsdevices/releases/tag/v6.2.0 https://github.com/pcdshub/pcdsdevices/releases/tag/v6.3.0  |
|  pcdsutils   | 0.10.0 | 0.11.0 |  https://github.com/pcdshub/pcdsutils/releases/tag/v0.11.0   |
|    pytmc     | 2.12.0 | 2.13.0 |    https://github.com/pcdshub/pytmc/releases/tag/v2.13.0     |
| transfocate  | 0.5.5  | 0.5.6  |  https://github.com/pcdshub/transfocate/releases/tag/v0.5.6  |
|    typhos    | 2.3.1  | 2.3.2  |    https://github.com/pcdshub/typhos/releases/tag/v2.3.2     |

SLAC Package Updates
--------------------

| Package |  Old   |  New   |                    Release Notes                     |
|:-------:|:------:|:------:|:----------------------------------------------------:|
|   pydm  | 1.16.1 | 1.16.3 | https://github.com/slaclab/pydm/releases/tag/v1.16.2 https://github.com/slaclab/pydm/releases/tag/v1.16.3 |

Added the Following Packages
----------------------------

- flask
- gammaray
- paramiko

Added the Following Dependencies
--------------------------------

- bcrypt (required by cryptography, paramiko, which are used in bluesky, bluesky-live, databroker, doctr, hutch-python, pcdsutils, pydm, sphinx, tiled, xarray)
- itsdangerous (required by flask, starlette, which are used in databroker, tiled)
- werkzeug (required by flask)

### CODE REVIEW

File: deploy/deploy_all [hunk 1]
```diff
@@ -68,7 +68,9 @@ ERROR_HOSTS=""
 set +e
 while read -r host; do
   echo "Deploying on ${host}"
-  ssh -n "${host}" "${DEPLOY}${INNER_ARGS}"
+  echo "ssh -n ${host} ${DEPLOY} ${INNER_ARGS}"
```
Comment (klauer) [inline deploy/deploy_all:71]: Nitpick: `set -x` before and `set +x` after instead?

File: scripts/create_incremental_env.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+# Create a new conda env that follows the updated specs but is as similar
+# to the previous environment as possible.
+# The previous environment is by default the one present on the master
+# branch, but we can create an incremental environment starting at any
+# valid git reference.
+
+if [ -z "${1}" ]; then
+  echo "Usage: create_incremental_env.sh [relnum] [base] [git_ref]"
```
Comment (klauer) [inline scripts/create_incremental_env.sh:9]: The `git_ref` thing is a nice touch 👍

File: deploy/pcds_env_deploy [hunk 3]
```diff
@@ -57,6 +61,21 @@ while getopts 're:i:mt:dh' OPTION; do
...
+
+# Delete old first
+if [ -n "${REMOVE_OLD}" ]; then
+  export PCDS_CONDA_VER="pcds-5.4.1"
+  source /cds/group/pcds/pyps/conda/pcds_conda
+  REMOVE="${HERE}/clean_old_packs.py"
+  if [ -z "${CMD}" ]; then
+    python ${REMOVE} --delete
+  else
+    python ${REMOVE}
```
Comment (klauer) [inline deploy/pcds_env_deploy:74]: Putting my decoding of these blocks into words:
* `$CMD` is empty if in "deploy" (not dry-run) mode
* So in this dry-run `else` clause we use the default (no arguments required) dry-run mode of `clean_old_packs.py`

**Review by klauer — APPROVED** [review]
New scripts, new environment, tool updates all LGTM.
Good to see that the "incremental update" tooling wasn't an extreme amount of effort.

File: deploy/pcds_env_deploy (hunk 3 above)
Comment (ZLLentz) [reply deploy/pcds_env_deploy:74]: Yeah that's right, this is admittedly very messy

File: deploy/deploy_all (hunk 1 above)
Comment (ZLLentz) [reply deploy/deploy_all:71]: Good call

**Comment by ZLLentz** [comment]
The "full env" builds are starting to hit some sort of CI performance limit... I saw an "unable to allocate" memory error.

File: deploy/pcds_env_deploy [hunk 4]
```diff
@@ -57,6 +61,21 @@ while getopts 're:i:mt:dh' OPTION; do
   esac
 done
 
+HERE="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+
+# Delete old first
+if [ -n "${REMOVE_OLD}" ]; then
+  export PCDS_CONDA_VER="pcds-5.4.1"
```
Comment (tangkong) [inline deploy/pcds_env_deploy:68]: Is this something we have to change every time we deploy?

**Review by tangkong — APPROVED** [review]
Looks good to me.  👍

**Comment by ZLLentz** [comment]
It seems to take about 5.2GB of RAM to run the 3.9 full build on our linux servers, which is less than the 7.5GB allocated for the travis runner. So I'm pretty confused about the OOM failures here while also being a bit disturbed that the usage is so high.
The OOM occurs after the solver is done while trying to do the environment transaction so I'm not immediately sure why the other envs aren't effected (aside from py310 which clearly dies sooner)

File: deploy/pcds_env_deploy (hunk 4 above)
Comment (ZLLentz) [reply deploy/pcds_env_deploy:68]: I just picked a random py3 env to use to run the python script

**Comment by klauer** [comment]
Worth a try? https://docs.travis-ci.com/user/reference/overview/#vm-instance-size

**Comment by ZLLentz** [comment]
Maybe... it's not clear from the docs pages if we can do those without "credits" even with a concurrency plan. I'm thinking that instead we might want to pull back a bit and run fewer jobs overall, maybe even just the "current env" and "next env (incremental)" builds

**Pull request merged**